File: java/json-transform/src/test/java/co/nlighten/jsontransform/functions/TransformerFunctionNormalizeTest.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.BaseTest;
import org.junit.jupiter.api.Test;

public class TransformerFunctionNormalizeTest extends BaseTest {

    @Test
    void testInlineFunctionNormalize() {
        var strVal = "Tĥïŝ ĩš â fůňķŷ Šťŕĭńġ ａｂｃＡＢＣ…";
        assertTransformation(strVal, "$$normalize:$", "This is a funky String abcABC...");
        assertTransformation(strVal, "$$normalize(NFKD):$", "This is a funky String abcABC...");
        assertTransformation(strVal, "$$normalize(NFKD,NONE):$", "Tĥïŝ ĩš â fůňķŷ Šťŕĭńġ abcABC...");
        assertTransformation(strVal, "$$normalize(NFD):$", "This is a funky String ａｂｃＡＢＣ…");
        assertTransformation(strVal, "$$normalize(NFD,NONE):$", "Tĥïŝ ĩš â fůňķŷ Šťŕĭńġ ａｂｃＡＢＣ…");

        // removeMarks on NFC/NFKC does not affect because marks are composed (and gone) before replacement
        assertTransformation(strVal, "$$normalize(NFKC):$", "Tĥïŝ ĩš â fůňķŷ Šťŕĭńġ abcABC...");
        assertTransformation(strVal, "$$normalize(NFKC,NONE):$", "Tĥïŝ ĩš â fůňķŷ Šťŕĭńġ abcABC...");
        assertTransformation(strVal, "$$normalize(NFC):$", strVal);
        assertTransformation(strVal, "$$normalize(NFC,NONE):$", strVal);
    }

    @Test
    void unicodeTableBasicLatin() {
        var strVal = "!\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~";
        assertTransformation(strVal, "$$normalize:$", strVal);
    }
    @Test
    void unicodeTableLatinExtendedA() {
        var strVal = "ĀāĂăĄąĆćĈĉĊċČčĎďĐđĒēĔĕĖėĘęĚěĜĝĞğĠġĢģĤĥĦħĨĩĪīĬĭĮįİıĲĳĴĵĶķĸĹĺĻļĽľĿŀŁłŃńŅņŇňŉŊŋŌōŎŏŐőŒœŔŕŖŗŘřŚśŜŝŞşŠšŢţŤťŦŧŨũŪūŬŭŮůŰűŲųŴŵŶŷŸŹźŻżŽžſ";
        var expect = "AaAaAaCcCcCcCcDdDdEeEeEeEeEeGgGgGgGgHhHhIiIiIiIiIıIJijJjKkĸLlLlLlLlLlNnNnNnnNnOoOoOoOEoeRrRrRrSsSsSsSsTtTtTtUuUuUuUuUuUuWwYyYZzZzZzs";
        assertTransformation(strVal, "$$normalize:$", expect);
    }

    @Test
    void unicodeTableLatinExtendedB() {
        var strVal = "ƀƁƂƃƄƅƆƇƈƉƊƋƌƍƎƏƐƑƒƓƔƕƖƗƘƙƚƛƜƝƞƟƠơƢƣƤƥƦƧƨƩƪƫƬƭƮƯưƱƲƳƴƵƶƷƸƹƺƻƼƽƾƿǀǁǂǃǄǅǆǇǈǉǊǋǌǍǎǏǐǑǒǓǔǕǖǗǘǙǚǛǜǝǞǟǠǡǢǣǤǥǦǧǨǩǪǫǬǭǮǯǰǱǲǳǴǵǶǷǸǹǺǻǼǽǾǿȀȁȂȃȄȅȆȇȈȉȊȋȌȍȎȏȐȑȒȓȔȕȖȗȘșȚțȜȝȞȟȠȡȢȣȤȥȦȧȨȩȪȫȬȭȮȯȰȱȲȳȴȵȶȷȸȹȺȻȼȽȾȿɀɁɂɃɄɅɆɇɈɉɊɋɌɍɎɏ";
        var expect = "bBƂƃbbCCcDDƋƌƍƎƏƐFfGƔhƖIKklƛƜƝƞƟOoƢƣPpRSsƩƪtTtƮUuƱƲYyZzƷƸƹƺƻƼƽƾƿǀǁǂǃDZDzdzLJLjljNJNjnjAaIiOoUuUuUuUuUuǝAaAaAEaeGgGgKkOoOoƷʒjDZDzdzGgHpNnAaAEaeOoAaAaEeEeIiIiOoOoRrRrUuUuSsTtȜȝHhȠdȢȣZzAaEeOoOoOoOoYylntȷȸȹACcLTszɁɂBUAEeJjQqRrYy";
        assertTransformation(strVal, "$$normalize:$", expect);
    }

    @Test
    void unicodeTableLatin1Supplement() {
        var strVal = "ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ";
        var expect = "AAAAAAAECEEEEIIIIDNOOOOO×OUUUUYÞßaaaaaaaeceeeeiiiiðnooooo÷ouuuuyþy";
        assertTransformation(strVal, "$$normalize:$", expect);
    }
    @Test
    void exceptionsAeOe() {
        var strVal = "ŒœÆæǢǣǼǽ";
        var expect = "OEoeAEaeAEaeAEae";
        assertTransformation(strVal, "$$normalize:$", expect);
    }
}

File: javascript/json-transform/src/__tests__/functions/TransformerFunctionNormalize.test.ts
import { describe, test } from "vitest";
import { assertTransformation } from "../BaseTransformationTest";

describe("TransformerFunctionNormalize", () => {
  test("inline", async () => {
    var strVal = "Tĥïŝ ĩš â fůňķŷ Šťŕĭńġ ａｂｃＡＢＣ…";
    await assertTransformation(strVal, "$$normalize:$", "This is a funky String abcABC...");
    await assertTransformation(strVal, "$$normalize(NFKD):$", "This is a funky String abcABC...");
    await assertTransformation(strVal, "$$normalize(NFKD,NONE):$", "Tĥïŝ ĩš â fůňķŷ Šťŕĭńġ abcABC...");
    await assertTransformation(strVal, "$$normalize(NFD):$", "This is a funky String ａｂｃＡＢＣ…");
    await assertTransformation(strVal, "$$normalize(NFD,NONE):$", "Tĥïŝ ĩš â fůňķŷ Šťŕĭńġ ａｂｃＡＢＣ…");

    // removeMarks on NFC/NFKC does not affect because marks are composed (and gone) before replacement
    await assertTransformation(strVal, "$$normalize(NFKC):$", "Tĥïŝ ĩš â fůňķŷ Šťŕĭńġ abcABC...");
    await assertTransformation(strVal, "$$normalize(NFKC,NONE):$", "Tĥïŝ ĩš â fůňķŷ Šťŕĭńġ abcABC...");
    await assertTransformation(strVal, "$$normalize(NFC):$", strVal);
    await assertTransformation(strVal, "$$normalize(NFC,NONE):$", strVal);
  });

  test("unicodeTableBasicLatin", async () => {
    var strVal = "!\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~";
    await assertTransformation(strVal, "$$normalize:$", strVal);
  });
  test("unicodeTableLatinExtendedA", async () => {
    var strVal =
      "ĀāĂăĄąĆćĈĉĊċČčĎďĐđĒēĔĕĖėĘęĚěĜĝĞğĠġĢģĤĥĦħĨĩĪīĬĭĮįİıĲĳĴĵĶķĸĹĺĻļĽľĿŀŁłŃńŅņŇňŉŊŋŌōŎŏŐőŒœŔŕŖŗŘřŚśŜŝŞşŠšŢţŤťŦŧŨũŪūŬŭŮůŰűŲųŴŵŶŷŸŹźŻżŽžſ";
    var expect =
      "AaAaAaCcCcCcCcDdDdEeEeEeEeEeGgGgGgGgHhHhIiIiIiIiIıIJijJjKkĸLlLlLlLlLlNnNnNnnNnOoOoOoOEoeRrRrRrSsSsSsSsTtTtTtUuUuUuUuUuUuWwYyYZzZzZzs";
    await assertTransformation(strVal, "$$normalize:$", expect);
  });

  test("unicodeTableLatinExtendedB", async () => {
    var strVal =
      "ƀƁƂƃƄƅƆƇƈƉƊƋƌƍƎƏƐƑƒƓƔƕƖƗƘƙƚƛƜƝƞƟƠơƢƣƤƥƦƧƨƩƪƫƬƭƮƯưƱƲƳƴƵƶƷƸƹƺƻƼƽƾƿǀǁǂǃǄǅǆǇǈǉǊǋǌǍǎǏǐǑǒǓǔǕǖǗǘǙǚǛǜǝǞǟǠǡǢǣǤǥǦǧǨǩǪǫǬǭǮǯǰǱǲǳǴǵǶǷǸǹǺǻǼǽǾǿȀȁȂȃȄȅȆȇȈȉȊȋȌȍȎȏȐȑȒȓȔȕȖȗȘșȚțȜȝȞȟȠȡȢȣȤȥȦȧȨȩȪȫȬȭȮȯȰȱȲȳȴȵȶȷȸȹȺȻȼȽȾȿɀɁɂɃɄɅɆɇɈɉɊɋɌɍɎɏ";
    var expect =
      "bBƂƃbbCCcDDƋƌƍƎƏƐFfGƔhƖIKklƛƜƝƞƟOoƢƣPpRSsƩƪtTtƮUuƱƲYyZzƷƸƹƺƻƼƽƾƿǀǁǂǃDZDzdzLJLjljNJNjnjAaIiOoUuUuUuUuUuǝAaAaAEaeGgGgKkOoOoƷʒjDZDzdzGgHpNnAaAEaeOoAaAaEeEeIiIiOoOoRrRrUuUuSsTtȜȝHhȠdȢȣZzAaEeOoOoOoOoYylntȷȸȹACcLTszɁɂBUAEeJjQqRrYy";
    await assertTransformation(strVal, "$$normalize:$", expect);
  });

  test("unicodeTableLatin1Supplement", async () => {
    var strVal = "ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ";
    var expect = "AAAAAAAECEEEEIIIIDNOOOOO×OUUUUYÞßaaaaaaaeceeeeiiiiðnooooo÷ouuuuyþy";
    await assertTransformation(strVal, "$$normalize:$", expect);
  });
  test("exceptionsAeOe", async () => {
    var strVal = "ŒœÆæǢǣǼǽ";
    var expect = "OEoeAEaeAEaeAEae";
    await assertTransformation(strVal, "$$normalize:$", expect);
  });
});
